[FLINK-36862][table] Implement additional TO_TIMESTAMP_LTZ() functions

yiyutian1 opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

The Flink TO_TIMESTAMP_LTZ is lacking some behaviors, compared with other vendors. eg. https://docs.snowflake.com/en/sql-reference/functions/to_timestamp

This PR adds the following additional functions for TO_TIMESTAMP_LTZ. 
```
TO_TIMESTAMP_LTZ(string1) -- Parse timestamp with default format 
TO_TIMESTAMP_LTZ(string1, string2) -- Parse timestamp with custom format 
TO_TIMESTAMP_LTZ(string1, string2, string3) -- Parse timestamp with format and timezone 
TO_TIMESTAMP_LTZ(numeric) -- For epoch milliseconds
```

## Brief change log
  - Add the additional overloading functions for TO_TIMESTAMP_LTZ.
  - Add java test cases. 
  - Make sure the current scala tests pass.


## Verifying this change

This change is already covered by existing tests, `TemporalTypesTest.scala`.
Added additional tests in `TimeFunctionsITCase.java`. 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: ( no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (documented in sql_functions.yml)

davidradl commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/TimeFunctionsITCase.java:
```diff
@@ -37,23 +38,26 @@
-                        temporalOverlapsTestCases(),
-                        ceilTestCases(),
-                        floorTestCases())
+//                        extractTestCases(),
```
why are we leaving in comments?

yiyutian1 replied:
Modified

davidradl commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:
```diff
@@ -2332,14 +2332,24 @@ ANY, and(logical(LogicalTypeRoot.BOOLEAN), LITERAL)
             BuiltInFunctionDefinition.newBuilder()
-                    .name("toTimestampLtz")
-                    .sqlName("TO_TIMESTAMP_LTZ")
+                    .name("TO_TIMESTAMP_LTZ")
```
looks like we are changing the name of an existing function - I assume this will impact migration. Don't we need to deprecate the old function and add the new one with the new name? And then put thjis in the documentation

yiyutian1 replied:
Hi @davidradl , it shouldn't impact the migration, because we are rewriting scala code into java, and the existing tests passed just fine.

yiyutian1 replied:
Hi @snuyanzin , gave it a try and spent much time on it. Still, for some reason, if I revert the two names back, when I run my new java tests, I got errors like 
`SQL validation failed. From line 1, column 8 to line 1, column 81: Cannot apply 'TO_TIMESTAMP_LTZ' to arguments of type 'TO_TIMESTAMP_LTZ(<CHAR(25)>, <CHAR(25)>)'. Supported form(s): 'TO_TIMESTAMP_LTZ(<NUMERIC>, <INTEGER>)'`

Could you shed some light on this? It seems like the SQL validation doesn't pick up my change.

yiyutian1 replied:
Reverted the change and made all existing tests pass.

yiyutian1 replied:
After some research, the new Java stack only has one "name", eg. [here](https://github.com/apache/flink/blob/01a2b28160f70a12d7200790acde53a9e910339a/flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java#L237). 
We are not changing the api name, it recognizes the api name [here](https://github.com/apache/flink/pull/25763/files#diff-d2ce68433255086dfdc898b9fd6636186044115a1280d39d6d04ebfd99933e94R377-R422).

davidradl commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ToTimestampLtzFunction.java:
```diff
@@ -0,0 +1,111 @@
+            return null;
+        }
+        if (precision == 0) {
+            return TimestampData.fromEpochMillis(epoch * 1000L);
```
Timestamp Data seems to have support for nanoseconds. I suggest tests with nanoseconds.

yiyutian1 replied:
Use the built-in function that the existing TO_TIMESTAMP_LTZ function depends on, which takes care of the nanoseconds.

davidradl commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ToTimestampLtzFunction.java:
```diff
@@ -0,0 +1,111 @@
+
+    // Parse timestamp with custom format
+    public TimestampData eval(String timestamp, String format) {
+        if (timestamp == null) {
```
NIT: could we just call the eval with the most arguments and check for null there? Instead of in each eval method.

yiyutian1 replied:
Hi @davidradl , thanks for the feedback, but for scalar functions we need to have them separate, aka 1 eval for 1 logic.

yiyutian1 commented on flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/expressions/TemporalTypesTest.scala:
```diff
@@ -1140,41 +1140,41 @@ class TemporalTypesTest extends ExpressionTestBase {
     tableConfig.setLocalTimeZone(ZoneId.of("Asia/Shanghai"))
 
     // INT -> TIMESTAMP_LTZ
-    testAllApis(toTimestampLtz(100, 0), "TO_TIMESTAMP_LTZ(100, 0)", "1970-01-01 08:01:40.000")
```
Converting the tests to expected behavior.

snuyanzin replied:
what is the Flink behavior before this PR and after?

For instance if we run a query from flink sql client?

yiyutian1 replied:
Synced with Sergey, this has always been the behavior for Flink. Therefore, I will revert these changes and make my new function align with this existing behavior.

snuyanzin commented on docs/data/sql_functions.yml:
```diff
@@ -679,9 +679,15 @@ temporal:
+  - sql: TO_TIMESTAMP_LTZ(numeric[, precision])
     table: toTimestampLtz(NUMERIC, PRECISION)
-    description: "Converts a epoch seconds or epoch milliseconds to a TIMESTAMP_LTZ, the valid precision is 0 or 3, the 0 represents TO_TIMESTAMP_LTZ(epochSeconds, 0), the 3 represents TO_TIMESTAMP_LTZ(epochMilliseconds, 3)."
+    description: "Converts a epoch seconds or epoch milliseconds to a TIMESTAMP_LTZ, the valid precision is 0 or 3, the 0 represents TO_TIMESTAMP_LTZ(epochSeconds, 0), the 3 represents TO_TIMESTAMP_LTZ(epochMilliseconds, 3). If no precision is provided, the default precision is 3."
```
```suggestion
    description: "Converts an epoch seconds or epoch milliseconds to a TIMESTAMP_LTZ, the valid precision is 0 or 3, the 0 represents TO_TIMESTAMP_LTZ(epochSeconds, 0), the 3 represents TO_TIMESTAMP_LTZ(epochMilliseconds, 3). If no precision is provided, the default precision is 3."
```

snuyanzin commented on docs/data/sql_functions.yml:
```diff
@@ -679,9 +679,15 @@ temporal:
     table: toDate(STRING1[, STRING2])
     description: Converts a date string string1 with format string2 (by default 'yyyy-MM-dd') to a date.
-  - sql: TO_TIMESTAMP_LTZ(numeric, precision)
+  - sql: TO_TIMESTAMP_LTZ(numeric[, precision])
```
Each time there is any update in doc, Chinese version also should be updated in same way.
It could be updated with same content in English however have to be kept in sync

yiyutian1 replied:
added doc to `sql_functions_zh.yml`

snuyanzin commented on docs/data/sql_functions.yml:
```diff
@@ -679,9 +679,15 @@ temporal:
+    description: "Converts a timestamp string string1 with format string2 (by default 'yyyy-MM-dd HH:mm:ss.SSS') to a TIMESTAMP_LTZ".
+  - sql: TO_TIMESTAMP_TZ(string1, string2, string3)
+    table: toTimestampLtz(STRING1, STRING2, STRING3)
+    description: "Converts a timestamp string string1 with format string2 to a TIMESTAMP_TZ in time zone string3."
```
What is `TIMESTAMP_TZ` ?
should it be `TIMESTAMP_LTZ`

snuyanzin commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ToTimestampLtzFunction.java:
```diff
@@ -0,0 +1,165 @@
+/**
+ * Implementation of {@link BuiltInFunctionDefinitions#TO_TIMESTAMP_LTZ}.
+ *
+ * <p>Supported function signatures: TO_TIMESTAMP_LTZ(numeric, precision) ->
+ * TIMESTAMP_LTZ(precision) TO_TIMESTAMP_LTZ(string) -> TIMESTAMP_LTZ(3) TO_TIMESTAMP_LTZ(string,
+ * format) -> TIMESTAMP_LTZ(3) TO_TIMESTAMP_LTZ(string, format, timezone) -> TIMESTAMP_LTZ(3)
+ * TO_TIMESTAMP_LTZ(numeric) -> TIMESTAMP_LTZ(3)
```
it's better to use `<pre>` tags for such things like here
https://github.com/apache/flink/blob/1580b4dd53cf023a201f12d822df0f9c1fbe241d/flink-test-utils-parent/flink-migration-test-utils/src/main/java/org/apache/flink/test/util/MigrationTest.java#L39-L42

snuyanzin commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ToTimestampLtzFunction.java:
```diff
@@ -0,0 +1,165 @@
+        super(BuiltInFunctionDefinitions.TO_TIMESTAMP_LTZ, context);
+    }
+
+    public TimestampData eval(Integer epoch, Integer precision) {
```
if it can return `null` please annotate it with  `@Nullable`
like https://github.com/apache/flink/blob/1580b4dd53cf023a201f12d822df0f9c1fbe241d/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/IfNullFunction.java#L35

snuyanzin commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ToTimestampLtzFunction.java:
```diff
@@ -0,0 +1,165 @@
+        return DateTimeUtils.toTimestampData(epoch, precision);
+    }
+
+    public TimestampData eval(Double epoch, Integer precision) {
```
This I didn't get, why do we have anything with floating point here?
Can you please clarify
first what does fractional epoch actually means? Does any vendor support it?
Second floating points have issues with exact data

yiyutian1 replied:
Hi Sergey, these are for existing test cases like:
https://github.com/apache/flink/blob/2d9fb799ed6647898ca353163723d3484dabfb6c/flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/expressions/TemporalTypesTest.scala#L1147

yiyutian1 replied:
Hi Sergey, they are for existing tests like:
https://github.com/apache/flink/blob/2d9fb799ed6647898ca353163723d3484dabfb6c/flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/expressions/TemporalTypesTest.scala#L1159

snuyanzin commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ToTimestampLtzFunction.java:
```diff
@@ -0,0 +1,165 @@
+        return DateTimeUtils.toTimestampData(epoch, precision);
+    }
+
+    public TimestampData eval(Float value, Integer precision) {
+        if (value == null || precision == null) {
+            return null;
+        }
+        return DateTimeUtils.toTimestampData(value.longValue(), precision);
+    }
+
+    public TimestampData eval(Byte value, Integer precision) {
+        if (value == null || precision == null) {
+            return null;
+        }
+        return DateTimeUtils.toTimestampData(value.longValue(), precision);
+    }
+
+    public TimestampData eval(DecimalData epoch, Integer precision) {
+        if (epoch == null || precision == null) {
+            return null;
+        }
+
+        return DateTimeUtils.toTimestampData(epoch, precision);
+    }
```
We need only numeric (exact numeric)

yiyutian1 replied:
same as above

snuyanzin commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ToTimestampLtzTypeStrategy.java:
```diff
@@ -23,19 +23,54 @@
+            return Optional.of(DataTypes.TIMESTAMP_LTZ(DEFAULT_PRECISION));
+        } else if (argCount == 2) {
+            LogicalTypeRoot firstArgTypeRoot = argumentTypes.get(0).getLogicalType().getTypeRoot();
+            boolean isFirstArgNumeric =
+                    firstArgTypeRoot == LogicalTypeRoot.TINYINT
+                            || firstArgTypeRoot == LogicalTypeRoot.SMALLINT
+                            || firstArgTypeRoot == LogicalTypeRoot.INTEGER
+                            || firstArgTypeRoot == LogicalTypeRoot.BIGINT
+                            || firstArgTypeRoot == LogicalTypeRoot.FLOAT
+                            || firstArgTypeRoot == LogicalTypeRoot.DOUBLE
+                            || firstArgTypeRoot == LogicalTypeRoot.DECIMAL;
+            // TO_TIMESTAMP_LTZ(numeric, precision)
```
can we replace it with check against logicalTypeFamily?
like here e.g.
https://github.com/apache/flink/blob/1580b4dd53cf023a201f12d822df0f9c1fbe241d/flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/CallSyntaxUtils.java#L106

snuyanzin commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ToTimestampLtzTypeStrategy.java:
```diff
@@ -23,19 +23,54 @@
-        if (callContext.isArgumentLiteral(1)) {
-            final int precision = callContext.getArgumentValue(1, Integer.class).get();
-            return Optional.of(DataTypes.TIMESTAMP_LTZ(precision));
+        List<DataType> argumentTypes = callContext.getArgumentDataTypes();
```
can you add tests for this logic
i guess somewhere in `org.apache.flink.table.types.inference` package
for instance `TypeStrategiesTest` or other class if other fits better

snuyanzin commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/DateTimeUtils.java:
```diff
@@ -386,6 +406,18 @@ public static TimestampData toTimestampData(DecimalData v, int precision) {
         }
     }
 
+    public static TimestampData toTimestampData(long epoch) {
+        return toTimestampData(epoch, 3);
+    }
+
+    public static TimestampData toTimestampData(double epoch) {
+        return toTimestampData(epoch, 3);
+    }
+
+    public static TimestampData toTimestampData(DecimalData epoch) {
+        return toTimestampData(epoch, 3);
+    }
+
```
from any user outside of `TO_TIMESTAMP_LTZ` 3 here is a magic number.
I would suggest to have such calls only in your toTimestampLtzFunction class and don't put them here

snuyanzin commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ToTimestampLtzFunction.java:
```diff
@@ -0,0 +1,165 @@
+        return DateTimeUtils.parseTimestampData(timestamp.toString(), format.toString());
+    }
+
+    // Parse timestamp with format and timezone
```
isn't name + args not self-sufficient?

snuyanzin commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ToTimestampLtzFunction.java:
```diff
@@ -0,0 +1,165 @@
+        return eval(epoch, 3);
+    }
+
+    public TimestampData eval(Long epoch) {
+        if (epoch == null) {
+            return null;
+        }
+
+        return eval(epoch, 3);
+    }
+
+    public TimestampData eval(Float epoch) {
+        if (epoch == null) {
+            return null;
+        }
+
+        return eval(epoch, 3);
+    }
```
Can not we replace a number of these duplicating methods with something like
```java
    public TimestampData eval(Number epoch) {
        if (epoch == null) {
            return null;
        }

        return eval(epoch, 3);
    }
```

yiyutian1 replied:
Address here: https://github.com/apache/flink/pull/25763/commits/533532547fc7ccf84e3ec6ee163597d3d61afcc2#r1886090826

snuyanzin commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ToTimestampLtzFunction.java:
```diff
@@ -0,0 +1,165 @@
+            return null;
+        }
+
+        return eval(epoch, 3);
```
IIRC there was somewhere a constant for this number
or can we introduce one here?

yiyutian1 commented:
@flinkbot run azure

davidradl commented:
Reviewed by Chi on 12/12/24 Group to test and/or review outside of the meeting (Nic Townsend from IBM has some insight he will add on this)

snuyanzin commented on flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/expressions/TemporalTypesTest.scala:
```diff
@@ -1158,7 +1157,7 @@ class TemporalTypesTest extends ExpressionTestBase {
       toTimestampLtz(100.01.cast(DataTypes.FLOAT()), 0),
       "TO_TIMESTAMP_LTZ(CAST(100.01 AS FLOAT), 0)",
-      "1970-01-01 08:01:40.010")
+      "1970-01-01 08:01:40.000")
```
why is it like that?
why are we missing millis now?

yiyutian1 replied:
Hi @snuyanzin, the 010 is the original test case. I reverted it back to 010 and the test passes now. 
I added a precision with 3 in the new Java test.

yiyutian1 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:
```diff
@@ -2332,7 +2332,8 @@ ANY, and(logical(LogicalTypeRoot.BOOLEAN), LITERAL)
             BuiltInFunctionDefinition.newBuilder()
-                    .name("TO_TIMESTAMP_LTZ")
+                    .name("toTimestampLtz")
+                    .sqlName("TO_TIMESTAMP_LTZ")
```
reverted the change of function name. Added necessary logic register the function in this commit.

yiyutian1 commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/TimeFunctionsITCase.java:
```diff
@@ -783,8 +790,15 @@ private Stream<TestSetSpec> toTimestampLtzTestCases() {
-                                TIMESTAMP_LTZ(0).nullable())
+                                null,
+                                TIMESTAMP_LTZ(3).nullable())
+                        .testResult(
```
To address this comment, added a FLOAT case with precision 3.

yiyutian1 commented on flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/expressions/TemporalTypesTest.scala:
```diff
@@ -1140,41 +1140,41 @@ class TemporalTypesTest extends ExpressionTestBase {
       toTimestampLtz(100.01.cast(DataTypes.FLOAT()), 0),
       "TO_TIMESTAMP_LTZ(CAST(100.01 AS FLOAT), 0)",
-      "1970-01-01 08:01:40")
+      "1970-01-01 08:01:40.010")
```
Reverted to the original test case.

yiyutian1 commented on flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/expressions/TemporalTypesTest.scala:
```diff
@@ -1308,29 +1303,42 @@ class TemporalTypesTest extends ExpressionTestBase {
-      "Cannot apply 'TO_TIMESTAMP_LTZ' to arguments of type" +
-        " 'TO_TIMESTAMP_LTZ(<CHAR(16)>, <INTEGER>)'. Supported form(s):" +
-        " 'TO_TIMESTAMP_LTZ(<NUMERIC>, <INTEGER>)'",
+      "SQL validation failed. From line 1, column 8 to line 1, column 46: Cannot apply " +
```
Need to change these existing scala tests, because I changed the function signature. The change reflects the expected function behaviors.

yiyutian1 commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ToTimestampLtzFunction.java:
```diff
@@ -92,74 +121,72 @@ public TimestampData eval(Integer epoch) {
             return null;
         }
 
-        return eval(epoch, 3);
```
Replying to this comment:
This is not possible, because we handle exact numeric and fractional data differently. Therefore, we can’t use the umbrella Number for them. 
I keep them as is to be consistent with existing logic. 
@snuyanzin , if you have a better way, please shed some lights here. 


> Can not we replace a number of these duplicating methods with something like
> 
>     public TimestampData eval(Number epoch) {
>         if (epoch == null) {
>             return null;
>         }
> 
>         return eval(epoch, 3);
>     }

yiyutian1 replied:
Resolved.

yiyutian1 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/DateTimeUtils.java:
```diff
@@ -385,6 +388,18 @@ public static TimestampData toTimestampData(int v, int precision) {
         }
     }
 
```
To address this comment, I replaced 3 with a constant. 
After giving it some thought, I believe it's better to keep these functions in DateTimeUtils.java rather than moving them to toTimestampLtzFunction, since toTimestampLtzFunction already depends on DateTimeUtils.java as a library. What are your thoughts on this? @snuyanzin 

> from any user outside of TO_TIMESTAMP_LTZ 3 here is a magic number.
> I would suggest to have such calls only in your toTimestampLtzFunction class and don't put them here

snuyanzin replied:
if you move them to toTimestampLtzFunction you can encapsulate the logic like mentioned in a comment https://github.com/apache/flink/pull/25763#discussion_r1890023415

yiyutian1 replied:
Moved this to the toTimestampLtzFunction.

yiyutian1 commented:
> Reviewed by Chi on 12/12/24 Group to test and/or review outside of the meeting (Nic Townsend from IBM has some insight he will add on this)

Hi @davidradl, I appreciate your involvement with this PR. As a new member to this community, this means a lot to me! 
However, this function is currently a blocker for a user at my company, Confluent. 

I've also just submitted another commit to address the previous comments, added tests, and made other updates, which may have rendered some of your current feedback outdated.

Given the time constraints, can I proceed with the Flink committer's approval, and address your comments in a separate PR?

snuyanzin commented on docs/data/sql_functions.yml:
```diff
@@ -679,9 +679,15 @@ temporal:
     table: toTimestampLtz(NUMERIC, PRECISION)
-    description: "Converts a epoch seconds or epoch milliseconds to a TIMESTAMP_LTZ, the valid precision is 0 or 3, the 0 represents TO_TIMESTAMP_LTZ(epochSeconds, 0), the 3 represents TO_TIMESTAMP_LTZ(epochMilliseconds, 3)."
+    description: Converts an epoch seconds or epoch milliseconds to a TIMESTAMP_LTZ, the valid precision is 0 or 3, the 0 represents TO_TIMESTAMP_LTZ(epochSeconds, 0), the 3 represents TO_TIMESTAMP_LTZ(epochMilliseconds, 3). If no precision is provided, the default precision is 3.
+  - sql: TO_TIMESTAMP_LTZ(string1[, string2])
+    table: toTimestampLtz(STRING1[, STRING2])
+    description: Converts a timestamp string string1 with format string2 (by default 'yyyy-MM-dd HH:mm:ss.SSS') to a TIMESTAMP_LTZ.
+  - sql: TO_TIMESTAMP_LTZ(string1, string2, string3)
+    table: toTimestampLtz(STRING1, STRING2, STRING3)
+    description: Converts a timestamp string string1 with format string2 to a TIMESTAMP_LTZ in time zone string3.
```
can we merge into one like
```yml
- sql: TO_TIMESTAMP_LTZ(string1[, string2[, string3]])
  table: toTimestampLtz(STRING1[, STRING2[, STRING3]])
```
?

snuyanzin commented on flink-python/pyflink/table/expressions.py:
```diff
@@ -306,19 +306,38 @@ def to_timestamp(timestamp_str: Union[str, Expression[str]],
-    3 means the numericEpochTime is in millisecond.
+    Supported signatures:
+    1. to_timestamp_ltz(numeric) -> timestamp_ltz
+    2. to_timestamp_ltz(numeric, precision) -> timestamp_ltz
+    3. to_timestamp_ltz(string) -> timestamp_ltz
+    4. to_timestamp_ltz(string, format) -> timestamp_ltz
+    5. to_timestamp_ltz(string, format, timezone) -> timestamp_ltz
```
IIRC in signatures we need to put types rather than names of vars

yiyutian1 replied:
Modified.

snuyanzin commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ToTimestampLtzFunction.java:
```diff
@@ -0,0 +1,192 @@
+    }
+
+    public @Nullable TimestampData eval(Integer epoch, Integer precision) {
+        if (epoch == null || precision == null) {
+            return null;
+        }
+
+        return DateTimeUtils.toTimestampData(epoch, precision);
+    }
+
+    public @Nullable TimestampData eval(Long epoch, Integer precision) {
+        if (epoch == null || precision == null) {
+            return null;
+        }
+
+        return DateTimeUtils.toTimestampData(epoch, precision);
+    }
+
+    public @Nullable TimestampData eval(Double epoch, Integer precision) {
+        if (epoch == null || precision == null) {
+            return null;
+        }
+
+        return DateTimeUtils.toTimestampData(epoch, precision);
+    }
+
+    public @Nullable TimestampData eval(Float value, Integer precision) {
+        if (value == null || precision == null) {
+            return null;
+        }
+        return DateTimeUtils.toTimestampData(value.doubleValue(), precision);
+    }
+
+    public @Nullable TimestampData eval(Byte value, Integer precision) {
+        if (value == null || precision == null) {
+            return null;
+        }
+        return DateTimeUtils.toTimestampData(value.longValue(), precision);
+    }
+
```
how about replacing this with something like
```java
    public @Nullable TimestampData eval(Number value, Integer precision) {
        if (value == null || precision == null) {
            return null;
        }
        if (value instanceof Float || value instanceof Double) {
            return DateTimeUtils.toTimestampData(value.doubleValue(), precision);
        }
        return DateTimeUtils.toTimestampData(value.longValue(), precision);
    }
```

yiyutian1 replied:
Modified.

snuyanzin commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/DateTimeUtils.java:
```diff
@@ -365,6 +368,38 @@ public static TimestampData toTimestampData(double v, int precision) {
+        }
+    }
+
+    public static TimestampData toTimestampData(long epoch) {
+        return toTimestampData(epoch, DEFAULT_PRECISION);
+    }
+
+    public static TimestampData toTimestampData(double epoch) {
+        return toTimestampData(epoch, DEFAULT_PRECISION);
+    }
+
+    public static TimestampData toTimestampData(DecimalData epoch) {
+        return toTimestampData(epoch, DEFAULT_PRECISION);
+    }
```
Move this to function class

yiyutian1 replied:
Hi @snuyanzin , thanks for the comment! 

I need to register these functions in `BuiltInMethods.scala` so that they can be exposed API calls. If I move these methods to my function class, I will have to  `lookupMethod` my new function class there. Do you think that’s preferred than the current way? My reasoning for the current way is that `lookupMethod` only need to call the common `DataTimeUtils.java` library class.

snuyanzin replied:
i wonder why do we still have anything in classes under `flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/` for this function?
If we want to move to new stack then there should be nothing.
IIRC there are only functions written in old stack approach

yiyutian1 replied:
Hi @snuyanzin , you asked a great question. I spent quite some time on this, and I think I figured it out. 

In this ticket, we aim to ensure that the existing Scala tests pass to confirm that the function's behavior remains unchanged, therefore we still depend on some scala auto-generated code, which requires modifications to the Scala folder. If I don't modify these folders, my new java tests fail because they can't pick up the new functions.

I attempted to use the new Java tech stack, but that caused the existing Scala tests to fail. 

Once the new functionality is out and stable, we should definitely complete the migration by removing the Scala tests and fully transition to Java, but now I think it's OK to keep it as is so we have the confidence that we don't break existing tests. What do you think?

snuyanzin replied:
Can you elaborate a bit more here what tests depend on it?

From my point of view during tests we should use either SQL or table api and both are public so ideally tests should not depend on internal implementation like old/new function stack

yiyutian1 replied:
Got rid of the scala tech stack in the new commit. (address 3rd round comment)

snuyanzin commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/DateTimeUtils.java:
```diff
@@ -365,6 +368,38 @@ public static TimestampData toTimestampData(double v, int precision) {
         }
     }
 
+    public static TimestampData toTimestampData(int v, int precision) {
+        switch (precision) {
+            case 0:
+                if (MIN_EPOCH_SECONDS <= v && v <= MAX_EPOCH_SECONDS) {
+                    return timestampDataFromEpochMills((v * MILLIS_PER_SECOND));
+                } else {
+                    return null;
+                }
+            case 3:
+                return timestampDataFromEpochMills(v);
+            default:
+                throw new TableException(
+                        "The precision value '"
+                                + precision
+                                + "' for function "
+                                + "TO_TIMESTAMP_LTZ(numeric, precision) is unsupported,"
+                                + " the supported value is '0' for second or '3' for millisecond.");
+        }
+    }
```
once feedback under comment https://github.com/apache/flink/pull/25763/files#r1889025297
is addressed, I think we can remove this method

snuyanzin commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ToTimestampLtzFunction.java:
```diff
@@ -0,0 +1,134 @@
+        return DateTimeUtils.toTimestampData(epoch, precision);
+    }
+
+    public @Nullable TimestampData eval(Number epoch) {
+        if (epoch == null) {
+            return null;
+        }
+        if (epoch instanceof Float || epoch instanceof Double) {
+            return DateTimeUtils.toTimestampData(epoch.doubleValue());
+        }
+        return DateTimeUtils.toTimestampData(epoch.longValue());
+    }
```
```suggestion
    public @Nullable TimestampData eval(Number epoch) {
        return eval(epoch, DEFAULT_PRECISION);
    }
```

great, it is becoming shorter
I would suggest one more step to reduce duplicated code by this
same could be applied for other method

yiyutian1 replied:
modified

nictownsend commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ToTimestampLtzFunction.java:
```diff
@@ -0,0 +1,128 @@
+        return DateTimeUtils.toTimestampData(epoch);
+    }
+
+    public @Nullable TimestampData eval(StringData timestamp) {
```
Possibly too much for this particular PR - but is it worth considering whether expanding the behaviour to support multiple default format?

For example, using https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatterBuilder.html to support multiple formats? E.g ISO + SQL as default matchers?

I have previously used something like https://www.baeldung.com/java-parsing-dates-many-formats#using-datetimeformatterbuilder when writing UDFs around parsing time to allow for multiple formats within a data source, or to allow for more flexible SQL re-use because I don't have to remember to swap between ISO and SQL or write date formats to cover different zone indcators.

nictownsend replied:
Counterpoint - Flink has multiple sources of data, where timestamp strings are not going to be SQL format.

It is frustrating to have to use `UNIX_TIMESTAMP(<date-time>,<format>`) with UTC format string (plus that implementation uses `SimpleDateFormat` so the format string is much more restricted than `DateTimeFormatter`).

`TO_TIMESTAMP` and `TO_TIMESTAMP_LTZ` functions with String input with "by default, SQL or ISO formats are supported" would be valuable.

Unless I've misunderstood as to your concerns about other vendors behaviour?

snuyanzin replied:
>It is frustrating to have to use UNIX_TIMESTAMP(<date-time>,<format>) with UTC format string (plus that implementation uses SimpleDateFormat so the format string is much more restricted than DateTimeFormatter).

>TO_TIMESTAMP and TO_TIMESTAMP_LTZ functions with String input with "by default, SQL or ISO formats are supported" would be valuable.

some default ISO formats should be supported as support was added long time ago. 
For other topics: looks like it is not related to this PR so feel free either ask a question in ML or create a jira issue for this and submit a PR (if you think it could/should be improved)

nictownsend replied:
Thank you, I wasn't sure if it was something to consider as part of these new functions. I will look more into the other default formats and use the appropriate channel.

snuyanzin commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ToTimestampLtzFunction.java:
```diff
@@ -0,0 +1,128 @@
+        return eval(epoch, DEFAULT_PRECISION);
+    }
+
+    public TimestampData eval(DecimalData epoch) {
+        if (epoch == null) {
+            return null;
+        }
+
+        return DateTimeUtils.toTimestampData(epoch);
+    }
```
same approach 
```suggestion
    public @Nullable TimestampData eval(DecimalData epoch) {
        return eval(epoch, DEFAULT_PRECISION);
    }
```

yiyutian1 replied:
modified

snuyanzin commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ToTimestampLtzFunction.java:
```diff
@@ -0,0 +1,128 @@
+ *
+ * <p>Example:
+ *
+ * <pre>{@code
+ * TO_TIMESTAMP_LTZ('2023-01-01 10:00:00')}  // Parses string using default format
+ * TO_TIMESTAMP_LTZ(1234567890123)}  // Converts epoch milliseconds
+ * TO_TIMESTAMP_LTZ(1234567890, 0)     // Converts epoch seconds
+ * TO_TIMESTAMP_LTZ(1234567890123, 3)  // Converts epoch milliseconds
+ * TO_TIMESTAMP_LTZ('2023-01-01 10:00:00')  // Parses string using default format
+ * TO_TIMESTAMP_LTZ('2023-01-01T10:00:00', 'yyyy-MM-dd\'T\'HH:mm:ss')} // Parses string using input format
+ * TO_TIMESTAMP_LTZ('2023-01-01 10:00:00', 'yyyy-MM-dd HH:mm:ss', 'UTC')} // Parses string using input format and timezone
+ * }</pre>
```
```suggestion
 * <pre>{@code
 * TO_TIMESTAMP_LTZ('2023-01-01 10:00:00')  // Parses string using default format
 * TO_TIMESTAMP_LTZ(1234567890123)  // Converts epoch milliseconds
 * TO_TIMESTAMP_LTZ(1234567890, 0)     // Converts epoch seconds
 * TO_TIMESTAMP_LTZ(1234567890123, 3)  // Converts epoch milliseconds
 * TO_TIMESTAMP_LTZ('2023-01-01 10:00:00')  // Parses string using default format
 * TO_TIMESTAMP_LTZ('2023-01-01T10:00:00', 'yyyy-MM-dd\'T\'HH:mm:ss') // Parses string using input format
 * TO_TIMESTAMP_LTZ('2023-01-01 10:00:00', 'yyyy-MM-dd HH:mm:ss', 'UTC') // Parses string using input format and timezone
 * }</pre>
```
There are several wrong curly brackets which break all the formatting
they should be removed

yiyutian1 replied:
modified

yiyutian1 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ToTimestampLtzTypeStrategy.java:
```diff
@@ -20,22 +20,79 @@
         }
-        return Optional.of(DataTypes.TIMESTAMP_LTZ(3));
+
+        return Optional.of(DataTypes.TIMESTAMP_LTZ(DEFAULT_PRECISION).nullable());
```
Synced with @jnh5y , because the function actually returns null, we should have the `.nullable()` here.

yiyutian1 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/DateTimeUtils.java:
```diff
@@ -421,6 +436,21 @@ public static TimestampData parseTimestampData(String dateStr, int precision, Ti
+        }
+
+        ZonedDateTime utcZoned = ts.toLocalDateTime().atZone(UTC_ZONE.toZoneId());
+        ZonedDateTime targetTime = utcZoned.withZoneSameInstant(ZoneId.of(timezone));
```
@nictownsend , thanks for the comment. After discussing with @snuyanzin and @jnh5y , we all agreed to modify this function to have the behavior you described.

yiyutian1 replied:
The conversion using UTC is not necessary. Can achieve the same result using:

# after getting the TimeStamp from parser:
ZonedDateTime zoneDate = ts.toLocalDateTime().atZone(ZoneId.of(timezone));
        return TimestampData.fromInstant(zoneDate.toInstant());

yiyutian1 replied:
@nictownsend could you resolve the `+-` status in the PR review?
I have addressed this comment.

yiyutian1 commented on flink-table/flink-table-api-scala/src/main/scala/org/apache/flink/table/api/ImplicitExpressionConversions.scala:
```diff
@@ -519,17 +519,6 @@ trait ImplicitExpressionConversions {
     Expressions.toTimestamp(timestampStr, format)
   }
 
-  /**
```
scala tech stack clean-up.

yiyutian1 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/DateTimeUtils.java:
```diff
@@ -436,21 +423,6 @@ public static TimestampData parseTimestampData(String dateStr, int precision, Ti
                         .toInstant());
     }
 
```
moved the logic to the Java function.

yiyutian1 commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/expressions/converter/DirectConvertRule.java:
```diff
@@ -285,9 +285,6 @@ void initNonDynamicFunctions() {
                 BuiltInFunctionDefinitions.UNIX_TIMESTAMP, FlinkSqlOperatorTable.UNIX_TIMESTAMP);
         definitionSqlOperatorHashMap.put(
                 BuiltInFunctionDefinitions.TO_DATE, FlinkSqlOperatorTable.TO_DATE);
-        definitionSqlOperatorHashMap.put(
```
old scala tech stack clean up.

yiyutian1 commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/FlinkSqlOperatorTable.java:
```diff
@@ -810,25 +810,6 @@ public SqlSyntax getSyntax() {
                             OperandTypes.family(SqlTypeFamily.CHARACTER, SqlTypeFamily.CHARACTER)),
                     SqlFunctionCategory.TIMEDATE);
 
```
old scala tech stack clean up.

yiyutian1 commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/BuiltInMethods.scala:
```diff
@@ -325,27 +325,6 @@ object BuiltInMethods {
     classOf[Long],
     classOf[TimeZone])
 
-  val LONG_TO_TIMESTAMP_LTZ_WITH_PRECISION =
```
old scala tech stack clean up.

yiyutian1 commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/FunctionGenerator.scala:
```diff
@@ -402,33 +402,6 @@ class FunctionGenerator private (tableConfig: ReadableConfig) {
 
   addSqlFunction(HASH_CODE, Seq(DECIMAL), new HashCodeCallGen())
 
-  INTEGRAL_TYPES.foreach(
```
old scala tech stack clean up.

yiyutian1 commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/StringCallGen.scala:
```diff
@@ -200,22 +200,6 @@ object StringCallGen {
             isCharacterString(operands(1).resultType) =>
         fallibleMethodGen(BuiltInMethods.STRING_TO_TIMESTAMP_WITH_FORMAT)
 
-      case TO_TIMESTAMP_LTZ if operands.size == 1 && isCharacterString(operands.head.resultType) =>
```
old scala tech stack clean up.

yiyutian1 commented on flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/expressions/TemporalTypesTest.scala:
```diff
@@ -1303,42 +1304,28 @@ class TemporalTypesTest extends ExpressionTestBase {
-        "'TO_TIMESTAMP_LTZ(<NUMERIC>)'\n" +
-        "'TO_TIMESTAMP_LTZ(<NUMERIC>, <INTEGER>)'\n" +
-        "'TO_TIMESTAMP_LTZ(<CHARACTER>, <CHARACTER>)'\n" +
-        "'TO_TIMESTAMP_LTZ(<CHARACTER>, <CHARACTER>, <CHARACTER>)'",
```
reflect exception message change after cleaning-up the legacy scala tech stack.

yiyutian1 commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/TimeFunctionsITCase.java:
```diff
@@ -926,7 +926,7 @@ private Stream<TestSetSpec> toTimestampLtzTestCases() {
                                         "yyyy-MM-dd HH:mm:ss",
                                         "Asia/Shanghai"),
                                 "TO_TIMESTAMP_LTZ('2023-01-01 00:00:00', 'yyyy-MM-dd HH:mm:ss', 'Asia/Shanghai')",
-                                LocalDateTime.of(2023, 1, 1, 8, 0, 0)
```
reflect behavior change, addressing[ this comment](https://github.com/apache/flink/pull/25763#discussion_r1891197006).

yiyutian1 commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ToTimestampLtzFunction.java:
```diff
@@ -89,46 +94,41 @@ public ToTimestampLtzFunction(SpecializedFunction.SpecializedContext context) {
 
     public @Nullable TimestampData eval(
-            StringData timestamp, StringData format, StringData timezone) {
-        if (timestamp == null || format == null || timezone == null) {
```
move the logic into from `DateTimeUtils.java` to `ToTimestampLtzFunction.java` to reduce redundant function calling.

yiyutian1 commented on flink-table/flink-table-api-scala/src/main/scala/org/apache/flink/table/api/ImplicitExpressionConversions.scala:
```diff
@@ -519,17 +519,6 @@ trait ImplicitExpressionConversions {
     Expressions.toTimestamp(timestampStr, format)
   }
 
```
Old scala tech stack clean-up.

yiyutian1 commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/batch/sql/TimeTravelTest.java:
```diff
@@ -133,7 +133,7 @@ void testTimeTravelWithUnsupportedExpression() {
                 .isInstanceOf(ValidationException.class)
                 .hasMessageContaining(
-                        "Unsupported time travel expression: TO_TIMESTAMP_LTZ(0, 3) for the expression can not be reduced to a constant by Flink.");
+                        "SQL validation failed. Unsupported time travel expression: TO_TIMESTAMP_LTZ(0, 3) for the expression can not be reduced to a constant by Flink.");
```
Modified tests to reflect the correct exception message, after cleaning-up the legacy scala tech stack.

snuyanzin commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ToTimestampLtzFunction.java:
```diff
@@ -0,0 +1,134 @@
+            return null;
+        }
+
+        ZonedDateTime zoneDate = ts.toLocalDateTime().atZone(ZoneId.of(timezone.toString()));
+        return TimestampData.fromInstant(zoneDate.toInstant());
+    }
```
what happen if there is invalid timezone?

yiyutian1 replied:
Did research on vendors' behaviors, when try parsing a timestamp with invalid timezone. 

MySQL:  returns NULL if any of the arguments are invalid, or if any of them are NULL.
Reference: https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_convert-tz

Oracle: return NULL if the input can't be parse to TIMESTAMP. 
Reference: https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/TO_TIMESTAMP_TZ.html

Flink returns NULL when timestamp string can't be parsed. 

I think we should return NULL here, considering there are other common vendors that have this behavior, and to be consistent to Flink's current behavior when we can't parse a timestamp string.

yiyutian1 replied:
I also added logic to return NULL when the **format** is not valid. 

Reference: 
Oracle: return NULL if the input can't be parse to TIMESTAMP.
Reference: https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/TO_TIMESTAMP_TZ.html

Flink returns NULL when timestamp string can't be parsed.

yiyutian1 commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/batch/sql/TimeTravelTest.java:
```diff
@@ -133,7 +133,7 @@ void testTimeTravelWithUnsupportedExpression() {
                 .isInstanceOf(ValidationException.class)
                 .hasMessageContaining(
-                        "Unsupported time travel expression: TO_TIMESTAMP_LTZ(0, 3) for the expression can not be reduced to a constant by Flink.");
+                        "SQL validation failed. Unsupported time travel expression: `TO_TIMESTAMP_LTZ`(0, 3) for the expression can not be reduced to a constant by Flink.");
```
Modified tests to reflect the correct exception message, after cleaning-up the legacy scala tech stack.

yiyutian1 commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/TimeFunctionsITCase.java:
```diff
@@ -947,6 +947,14 @@ private Stream<TestSetSpec> toTimestampLtzTestCases() {
                                         .atZone(ZoneId.of("America/Los_Angeles"))
                                         .toInstant(),
                                 TIMESTAMP_LTZ(3).nullable())
```
Address this comment: 
https://github.com/apache/flink/pull/25763#discussion_r1892889885

jnh5y replied:
Makes sense to me.

hanyuzheng7 replied:
lgtm.

yiyutian1 commented on docs/data/sql_functions_zh.yml:
```diff
@@ -805,11 +805,12 @@ temporal:
-      3 代表` TO_TIMESTAMP_LTZ(epochMilliseconds, 3)`。
+  - sql: TO_TIMESTAMP_LTZ(numeric[, precision])
+    table: toTimestampLtz(NUMERIC, PRECISION)
+    description: Converts an epoch seconds or epoch milliseconds to a TIMESTAMP_LTZ, the valid precision is 0 or 3, the 0 represents TO_TIMESTAMP_LTZ(epochSeconds, 0), the 3 represents TO_TIMESTAMP_LTZ(epochMilliseconds, 3). If no precision is provided, the default precision is 3.
```
Hi Peter, thanks for the comment! it's OK to keep it both in English at the time being. 
[reference](https://github.com/apache/flink/pull/25763#discussion_r1881834036)

hanyuzheng7 replied:
ok

yiyutian1 commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/DateTimeUtils.java:
```diff
@@ -424,8 +424,12 @@ public static TimestampData parseTimestampData(String dateStr, int precision, Ti
 
     public static TimestampData parseTimestampData(String dateStr, String format) {
-        DateTimeFormatter formatter = DATETIME_FORMATTER_CACHE.get(format);
-
```
Address comment. https://github.com/apache/flink/pull/25763#discussion_r1893222879

jnh5y replied:
Nice!  This also makes sense to catch the IllegalArgumentException / parse errors!

snuyanzin commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Expressions.java:
```diff
@@ -366,6 +366,61 @@ public static ApiExpression toTimestampLtz(Object numericEpochTime, Object preci
     }
 
+    /**
+     * * Converts the given time string with the specified format to {@link
```
```suggestion
     * Converts the given time string with the specified format to {@link
```

yiyutian1 replied:
modified.

snuyanzin commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ToTimestampLtzFunction.java:
```diff
@@ -0,0 +1,139 @@
+            return null;
+        }
+
+        TimestampData ts = parseTimestampData(dateStr.toString(), format.toString());
+        if (ts == null) {
+            return null;
+        }
+
+        try {
+            ZonedDateTime zoneDate = ts.toLocalDateTime().atZone(ZoneId.of(timezone.toString()));
+            return TimestampData.fromInstant(zoneDate.toInstant());
+        } catch (DateTimeException e) {
+            return null;
+        }
+    }
```
As I mentioned earlier (probably missed) Introduce a `parseTimestampData` method with 3 input args in `DateTimeUtils` for that
something like
```java
    public static TimestampData parseTimestampData(String dateStr, int precision, TimeZone timeZone)
            throws DateTimeException {
     
                  return TimestampData.fromInstant(
            fromTemporalAccessor(DateTimeFormatter.ofPattern(format).parse(dateStr), 3)
                    .atZone(ZoneId.of(timezone))
                    .toInstant());
    }
```
+ also be sure that it works for invalid format and invalid timezone

Do we have tests for such invalid inputs?

yiyutian1 replied:
Hi @snuyanzin This method is already in `DateTimeUtils.java`, I decided not to use it because it doesn't handle the exceptions well (eg, when we have invalid **timestamp** and **format**. for those cases, I return null here), and I re-wrote the function in the `ToTimestampLtzFunction.java` class. 

Also, for invalid format and timezone, I have them here: https://github.com/apache/flink/pull/25763/files/982b92f13fbb93b8e0c5244b30d4ae8bbbfa3a25#diff-7588925e682b7f5b4b2e41fdc4992ef0d9f246ce00696f783c36f993b8aeb7aeR950-R973

snuyanzin replied:
yep, thanks tests are there

about method call
here in this method there should just one call as below (same approach as for other methods in this class)
this is what I mean
let me know whether it is clear or not
```java
    public @Nullable TimestampData eval(
            StringData dateStr, StringData format, StringData timezone) {
        if (dateStr == null || format == null || timezone == null) {
            return null;
        }
        // call the method from DateTimeUtils
        parseTimestampData();
        }
```

yiyutian1 replied:
Hi @snuyanzin , thank you for clarifying!

I understand that, but that method doesn't handle the invalid cases the way we want here. When we have invalid **timestamp** and **format** and **zoneID**, here we want to return NULL, but that method just throws exception. That's why I decided to re-write and handle those invalid cases separately.


That method with the particular signature is never used, and I don't think we need to refer to it as source of truth. What do you think? 
<img width="905" alt="Screenshot 2024-12-20 at 8 31 03 AM" src="https://github.com/user-attachments/assets/9f72231e-12d8-446d-9613-e0bc6a6be8a0" />

yiyutian1 replied:
Gave it a try the way your recommended, I don’t think it’s essentially different from what I have, and it actually requires more code change. 

**You recommended way would need:** 
Step 1: Add this method to `DateTimeUtils.java`. 
```
public static TimestampData parseTimestampData(String dateStr, int precision, TimeZone timeZone)
            throws DateTimeException {
        return TimestampData.fromInstant(
                fromTemporalAccessor(DEFAULT_TIMESTAMP_FORMATTER.parse(dateStr), precision)
                        .atZone(timeZone.toZoneId())
                        .toInstant());
    }

```
Step 2: Call it in the `ToTimestampLtzFunction.java` function:
```
public @Nullable TimestampData eval(
            StringData dateStr, StringData format, StringData timezone) {
        if (dateStr == null || format == null || timezone == null) {
            return null;
        }

        try {
            return parseTimestampData(dateStr.toString(), format.toString(), ZoneId.of(timezone.toString()));
        } catch (DateTimeException e) {
            return null;
        }
}
```

`My way: `
In `ToTimestampLtzFunction.java`:
 ```
public @Nullable TimestampData eval(
            StringData dateStr, StringData format, StringData timezone) {
        if (dateStr == null || format == null || timezone == null) {
            return null;
        }

        TimestampData ts = parseTimestampData(dateStr.toString(), format.toString());
        if (ts == null) {
            return null;
        }

        try {
            ZonedDateTime zoneDate = ts.toLocalDateTime().atZone(ZoneId.of(timezone.toString()));
            return TimestampData.fromInstant(zoneDate.toInstant());
        } catch (DateTimeException e) {
            return null;
}
```
Again, I don't see a particular benefit in doing it one way over another, but I'm open for your feedback.

snuyanzin commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/TimeFunctionsITCase.java:
```diff
@@ -804,4 +813,163 @@ private Stream<TestSetSpec> floorTestCases() {
+                                1234,
+                                -100,
+                                null,
+                                DecimalDataUtils.castFrom(-Double.MAX_VALUE, 38, 18),
```
I don't see here tests for invalid input
like invalid timezone, invalid format
invalid something else...

yiyutian1 replied:
Hi @snuyanzin , I have the invalid test cases here, the invalid inputs are directly pased in.  https://github.com/apache/flink/pull/25763/files/982b92f13fbb93b8e0c5244b30d4ae8bbbfa3a25#diff-7588925e682b7f5b4b2e41fdc4992ef0d9f246ce00696f783c36f993b8aeb7aeR950-R973

Could you clarify what you mean in this comment?

yiyutian1 replied:
Resolved.

hanyuzheng7 commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ToTimestampLtzFunction.java:
```diff
@@ -0,0 +1,139 @@
+            ZonedDateTime zoneDate = ts.toLocalDateTime().atZone(ZoneId.of(timezone.toString()));
+            return TimestampData.fromInstant(zoneDate.toInstant());
+        } catch (DateTimeException e) {
+            return null;
```
LGTM, just a question, why we return null in catch block?

davidradl replied:
Good question - should we log this exception rather than just swallowing it?

yiyutian1 replied:
Good question, @davidradl and @hanyuzheng7 !

I explained the reasoning here: https://github.com/apache/flink/pull/25763#discussion_r1893222879
We return null here because it's the desired behavior.
Also because it's consistent to how we handle exceptions in DateTime, eg. [here](https://github.com/apache/flink/blob/7fa4f78e43bfd99a7e7a3d655a9a3641e4d7bc59/flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/DateTimeUtils.java#L447). 

Logging would be fine, but that also requires refactoring other functions to make it consistent, maybe in a separate effort.

snuyanzin replied:
No logging in the “hot path“ please
We are building a streaming/batch system that processes millions of events per minute
 Logging could fill a local TM disk or the log provider easily.

yiyutian1 replied:
Good to know the specific reasoning! TIL

snuyanzin commented:
There is still one case found showing regression

before PR
```sql
SELECT to_timestamp_ltz(123, CAST(NULL AS INTEGER));
```
before PR returns `null`

after PR it fails on validation

snuyanzin commented on docs/data/sql_functions.yml:
```diff
@@ -679,9 +679,12 @@ temporal:
+  - sql: TO_TIMESTAMP_LTZ(numeric[, precision])
     table: toTimestampLtz(NUMERIC, PRECISION)
-    description: "Converts a epoch seconds or epoch milliseconds to a TIMESTAMP_LTZ, the valid precision is 0 or 3, the 0 represents TO_TIMESTAMP_LTZ(epochSeconds, 0), the 3 represents TO_TIMESTAMP_LTZ(epochMilliseconds, 3)."
+    description: Converts an epoch seconds or epoch milliseconds to a TIMESTAMP_LTZ, the valid precision is 0 or 3, the 0 represents TO_TIMESTAMP_LTZ(epochSeconds, 0), the 3 represents TO_TIMESTAMP_LTZ(epochMilliseconds, 3). If no precision is provided, the default precision is 3.
+  - sql: TO_TIMESTAMP_LTZ(string1[, string2[, string3]])
+    table: toTimestampLtz(STRING1[, STRING2[, STRING3]])
+    description: Converts a timestamp string string1 with format string2 (by default 'yyyy-MM-dd HH:mm:ss.SSS') to a TIMESTAMP_LTZ in time zone string3. If no time zone is provided, the default time zone is the session time zone.
```
I think we should explicitly mention that any null input leads to null output like it is done for some other functions (especially array related)

yiyutian1 commented:
> Thanks for addressing feedback it looks we are approaching
> 
> There is at least one case found showing regression
> 
> ```sql
> SELECT to_timestamp_ltz(123, CAST(NULL AS INTEGER));
> ```
> 
> before PR returns `null`
> 
> after PR it fails on validation

such good catch! Fixed it in the new commit. thanks for the thorough checks.

yiyutian1 commented on flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/TimeFunctionsITCase.java:
```diff
@@ -853,18 +848,13 @@ private Stream<TestSetSpec> toTimestampLtzTestCases() {
                         .testResult(
-                                toTimestampLtz($("f5"), literal(3)),
-                                "TO_TIMESTAMP_LTZ(f5, 3)",
+                                toTimestampLtz($("f4"), literal(3)),
```
Moved one test case with null input to the end of the stream.

yiyutian1 commented:
@flinkbot run azure

yiyutian1 commented on docs/data/sql_functions.yml:
```diff
@@ -681,10 +681,10 @@ temporal:
   - sql: TO_TIMESTAMP_LTZ(string1[, string2[, string3]])
     table: toTimestampLtz(STRING1[, STRING2[, STRING3]])
-    description: Converts a timestamp string string1 with format string2 (by default 'yyyy-MM-dd HH:mm:ss.SSS') to a TIMESTAMP_LTZ in time zone string3. If no time zone is provided, the default time zone is the session time zone.
+    description: Converts a timestamp string string1 with format string2 (by default 'yyyy-MM-dd HH:mm:ss.SSS') in time zone string3 (by default 'UTC') to a TIMESTAMP_LTZ. If any input is null, the function will return null.
```
Adjust the description slightly, to be more clear.

snuyanzin reviewed (approved):
Thanks a lot for your efforts here
I tested locally (found one minor issue which doesn't relate to the changes)
nothing else from my side
LGTM


@nictownsend since you were the one who requested changes
do you have anything to add here?

Also similar question to
@davidradl , @jnh5y , @hanyuzheng7 


I'm asking since if no objections I'm going to merge by the end of this week

nictownsend reviewed (approved):
Approving my requested changes reqarding local date time and zones

davidradl commented on docs/data/sql_functions_zh.yml:
```diff
@@ -805,11 +805,12 @@ temporal:
+    description: Converts an epoch seconds or epoch milliseconds to a TIMESTAMP_LTZ, the valid precision is 0 or 3, the 0 represents TO_TIMESTAMP_LTZ(epochSeconds, 0), the 3 represents TO_TIMESTAMP_LTZ(epochMilliseconds, 3). If no precision is provided, the default precision is 3. If any input is null, the function will return null.
+  - sql: TO_TIMESTAMP_LTZ(string1[, string2[, string3]])
+    table: toTimestampLtz(STRING1[, STRING2[, STRING3]])
+    description: Converts a timestamp string string1 with format string2 (by default 'yyyy-MM-dd HH:mm:ss.SSS') in time zone string3 (by default 'UTC') to a TIMESTAMP_LTZ. If any input is null, the function will return null.
```
is the translation to chinese going to be tracked else where?

yiyutian1 replied:
Hi @davidradl, thanks for the comment. Will create a Jira for people to pick it up.

yiyutian1 replied:
Created Jira: https://issues.apache.org/jira/browse/FLINK-37027

davidradl commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Expressions.java:
```diff
@@ -366,6 +366,61 @@ public static ApiExpression toTimestampLtz(Object numericEpochTime, Object preci
+    /**
+     * Converts a timestamp to {@link DataTypes#TIMESTAMP_LTZ(int)}.
+     *
+     * <p>This method takes an object representing a timestamp and converts it to a TIMESTAMP_LTZ
```
nit: object -> String

yiyutian1 replied:
Addressed.

davidradl commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ToTimestampLtzTypeStrategy.java:
```diff
@@ -20,22 +20,79 @@
+                    || !isCharacterType(argumentTypes.get(2).getLogicalType().getTypeRoot())) {
+                throw new ValidationException(
+                        "Unsupported argument type. "
+                                + "When taking 3 arguments, TO_TIMESTAMP_LTZ requires all three arguments to be of type <CHARACTER>.");
```
what does <CHARACTER> mean - I assume we mean the Flink logical types <CHAR> or <VARCHAR> . I assume that STRING resolves to VARCHAR.

yiyutian1 replied:
Will modify.

yiyutian1 commented on flink-python/pyflink/table/tests/test_expression.py:
```diff
@@ -285,7 +285,14 @@ def test_expressions(self):
         self.assertEqual("toDate('2018-03-18')", str(to_date('2018-03-18')))
         self.assertEqual("toDate('2018-03-18', 'yyyy-MM-dd')",
                          str(to_date('2018-03-18', 'yyyy-MM-dd')))
-        self.assertEqual('toTimestampLtz(123, 0)', str(to_timestamp_ltz(123, 0)))
```
Hi @davidradl , I checked those negatives cases in Java tests. To reduce the python overhead in Flink, we usually don't duplicate tests in python tests. 

Also, Flink tests are very light-weight, and only check function **signatures**, not the function themselves. In other words, even a negative test won't return error here. If we add one here, it might lead to some confusion.

jnh5y reviewed (approved):
@snuyanzin I do not have any concerns.  I think this review has been thorough and covered things well!

hanyuzheng7 reviewed (approved):
lgtm.

yiyutian1 commented on flink-python/pyflink/table/expressions.py:
```diff
@@ -306,19 +306,38 @@ def to_timestamp(timestamp_str: Union[str, Expression[str]],
-    3 means the numericEpochTime is in millisecond.
+    Supported functions:
+    1. to_timestamp_ltz(Numeric) -> DataTypes.TIMESTAMP_LTZ
+    2. to_timestamp_ltz(Numeric, Integer) -> DataTypes.TIMESTAMP_LTZ
```
Addressed.

yiyutian1 commented:
Hi @davidradl, thank you for the thorough check! I appreciate it. I’ve addressed your comments in the last commit. 

Since the PR is functionally intact, thoroughly reviewed, and has been approved by a Flink committer, I propose we merge it to improve iteration velocity, especially due to [the upcoming feature freeze](https://lists.apache.org/thread/yk1t5n9m7992vhqrscdr7o0gm0x7z2gh) next week. This allows users to start benefiting from the changes without being blocked.
If you have further feedback of improvement, if the issue doesn’t block the functionality, I’m happy to create a separate Jira and address those in a follow-up PR. What do you think?

cc: @snuyanzin
